Build using `-pedantic`; in Autoconf maintainer mode

jbenden opened the pull request:
This patch series goes through the source code to enable builds, without warnings or errors, under the strictness of `-pedantic` compiler flag. Additionally, some of the shell scripts were fixed with `shellcheck`.

This patch series introduces the selection of a C++ standard (non-required) to assess the availability across the CI infrastructure.

This patch series fixes the great majority of outstanding PVS-Studio warnings.

Finally, and most important for other developers, is that we all must enable maintainer mode during our development; pass `--enable-maintainer-mode` to the `configure` script. This Autoconf flag is what enables the `-pedantic` warnings.

jbenden committed:
fix(ce-wpa): anonymous unions are a C11 extension

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden committed:
refactor: fix undefined behaviour

This removes the use of macros within strings. This is an undefined
usage. Instead, simply print all possible command-line arguments and
show an error message if an unsupported option is selected.

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden committed:
refactor: fix function callback types

- Callbacks should be passed as the address-of a function.
- Mark all thread routines with THREAD_ENTRY to make it very
  clear that these function run in a new thread.

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden committed:
airodump-ng: binary constants are a gnu extension

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden committed:
refactor: use c99-style

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden committed:
refactor: use of \e is a gnu extension

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden committed:
unit-tests: must have cast to clean compile

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden committed:
unit-tests: conversion to constant expr is a gcc extension

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden committed:
mark gcc specific: braced-group within expression

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden committed:
fix scanf formats expecting unsigned int

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden committed:
autotools: refactor maintainer mode

- Disable maintainer mode, by default. Now use the flag
  `--enable-maintainer-mode` when developing both autoconf
  and source code!
- When maintainer mode is enabled; add pedantic warnings.

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden committed:
fixed all shellcheck reported warnings

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden committed:
re-ran clang-format-3.8 over sources

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden committed:
remove loop unrolls

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden committed:
refactor(ce-wpa): spring clean formulas

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden committed:
aircrack-ng: fix incorrect usage of phread_mutex_trylock

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden committed:
autotools: Check for modern C++ support

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden committed:
autotools: reduce build noise

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden committed:
besside-ng: re-format source code

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden committed:
autotools: initial support for LTO builds

This patch series untangles the mixed assembly syntax, thereby
removing the need for `-masm=intel`.

Added autoconf flag to enable LTO: `--with-lto`.

As this is basic support (works for me! :P), it is not enabled
by default.

Signed-off-by: Joseph Benden <joe@benden.us>

aircrack-ng commented:
Could you document the new configure flag (`--enable-maintainer-mode`) in README.md and INSTALLING?

Could you update the case of PCRE to make it uppercase in besside-ng usage?

jbenden committed:
Updated the case of PCRE in usages and errors.

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden committed:
Document the --enable-maintainer-mode configure flag

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden commented:
On 8/23/20 5:15 PM, Aircrack-ng wrote:
>
> Could you document the new configure flag (|--enable-maintainer-mode|)
> in README.md and INSTALLING?
>
> Could you update the case of PCRE to make it uppercase in besside-ng
> usage?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/aircrack-ng/aircrack-ng/pull/2174#issuecomment-678845070>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAASAJ6UFUVACRCJRVI3QQDSCGWK7ANCNFSM4QD6KITA>.
>
I have just published additional commits which tackle these items.

Let me know if I missed something.

Suggestions, comments, and ideas are always welcomed,
-Joe

jbenden committed:
fix(lib/libac): add cache-line padding around fields

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden committed:
fix(airodump-ng): fix crash on 4K Linux console

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden committed:
simplify code; add validation of port number

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden committed:
add PVS-Studio suppressions for DbC macros

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden committed:
many type fixes

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden committed:
refactor: removal of duplicate code

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden committed:
refactor: general cleaning via PVS-Studio

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden committed:
refactor: remove possibility of NULL deref

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden committed:
refactor(ce-wpa): remove old-school memory checker

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden committed:
Bulk fixes for PVS-Studio warnings

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden committed:
reformat source code with clang-format-3.8

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden committed:
fix(wesside-ng): fix maths assumptions

Fix:

  assuming signed overflow does not occur when simplifying conditional to
  constant

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden commented:
@Mister-X- Would you mind reviewing this for a merge to the master branch?

jbenden committed:
general clean-up

- Mark unused function parameters as such.
- Type corrections.
- Remove unused variables.

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden committed:
easside-ng: fix code warning under -Wextra

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden committed:
autotools: increased developer warnings

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden committed:
fix(aircrack-ng): correctly signal cond variables

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden committed:
fix(libac): mark unused parameters

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden committed:
fix(osdep/freebsd_tap): use full buffer

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden committed:
fix type and initialization issues

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden committed:
add missing include header

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden committed:
fix(osdep/netbsd_tap): fix size of buffer

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden committed:
fix(tkiptun-ng): type fix

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden committed:
fix(airodump-ng): test explicitly against unsigned char

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden committed:
fix(libac): DragonFlyBSD uses special include for *_np

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden committed:
type cannot be negative

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden committed:
autotools: do not enable -Werror with icc

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden committed:
fix(aircrack-ng): fix return value

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden committed:
fix(airventriloquist-ng): change case expression to a block

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden committed:
autotools: enable -pedantic when GCC v4.9 or later

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden committed:
refactor: create a fallthrough macro

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden committed:
refactor(aircrack-ng): adjust inlining

- Thread entry functions must never be inlined.
- Mark certain functions as noinline, based on ICC optimization
  reports.

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden committed:
autotools: disable ICC remark reporting

This commit is to keep ICC from reporting remarks about inlining.

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden committed:
tests: remove non-standard preprocessor directive

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden committed:
unittests: change usage to prevent -pedantic error

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden committed:
autotools: do not complain about zero variadic macro args

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden committed:
appveyor: always save cache and do not package PRs

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden committed:
fix(aircrack-ng): validate user input for process count

- Fully validate the number of processes desired (`-p`);
  ensuring a selection is made from a fully valid range.
  Fixes (CID 1077260)
- Display a warning message if the number of processes
  exceed the number of CPUs.

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden committed:
fix(aireplay-ng): fix potential division-by-zero (CID 1077329)

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden committed:
fix(osdep/cygwin): suppress dlsym errors

See: https://stackoverflow.com/questions/36384195/how-to-correctly-assign-a-pointer-returned-by-dlsym-into-a-variable-of-function
Signed-off-by: Joseph Benden <joe@benden.us>

jbenden committed:
appveyor: enable maintainer mode

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden committed:
fix(build): fix compilation with -DNDEBUG

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden committed:
fix(osdep/cygwin_tap): fix mismatched signed comparisons

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden committed:
fix(osdep/airpcap): exclude header file from pedantic

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden committed:
fix unmarked fall-through branches

Signed-off-by: Joseph Benden <joe@benden.us>
